### PR TITLE
[SEDONA-606] Add ST_IsValidDetail

### PR DESCRIPTION
**Pull request by furqaankhan**
## Did you read the Contributor Guide?

- Yes, I have read the [Contributor Rules](https://sedona.apache.org/latest-snapshot/community/rule/) and [Contributor Development Guide](https://sedona.apache.org/latest-snapshot/community/develop/)

## Is this PR related to a JIRA ticket?

- Yes, the URL of the associated JIRA ticket is https://issues.apache.org/jira/browse/SEDONA-606. The PR name follows the format `[SEDONA-XXX] my subject`.

## What changes were proposed in this PR?

- Add ST_IsValidDetail

## How was this patch tested?

- Add new unit tests

## Did this PR include necessary documentation updates?

- Yes, I have updated the documentation.

### CODE REVIEW

**Comment by jiayuasu** [comment]
@furqaankhan Please create a PR on `sedona-snowflake-tester` to test snowflake functions too

File: common/src/main/java/org/apache/sedona/common/utils/ValidDetail.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    public ValidDetail(boolean valid, String reason, Geometry location) {
+        this.valid = valid;
+        this.reason = reason;
+        this.location = location;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("Valid: %b\nReason: %s\nLocation: %s", valid, reason, Functions.asWKT(location));
```
Comment (jiayuasu) [inline common/src/main/java/org/apache/sedona/common/utils/ValidDetail.java:37]: null check for this?

File: common/src/test/java/org/apache/sedona/common/FunctionsTest.java [hunk 2]
```diff
@@ -2861,6 +2862,44 @@ public void locateAlong() throws ParseException {
...
+        // Valid geometry
+        Geometry validGeom = GEOMETRY_FACTORY.createPolygon(coordArray(30, 10, 40, 40, 20, 40, 10, 20, 30, 10));
+        ValidDetail actualValidDetail = Functions.isValidDetail(validGeom);
+        ValidDetail expectedValidDetail = new ValidDetail(true, null, null);
+        assertTrue(expectedValidDetail.equals(actualValidDetail));
+
+        Integer OGC_SFS_VALIDITY = 0;
+        Integer ESRI_VALIDITY = 1;
+
+        actualValidDetail = Functions.isValidDetail(validGeom, OGC_SFS_VALIDITY);
```
Comment (jiayuasu) [inline common/src/test/java/org/apache/sedona/common/FunctionsTest.java:2876]: add a test case when any property is null?

File: docs/api/snowflake/vector-data/Function.md [hunk 3]
```diff
@@ -1486,6 +1486,35 @@ Output:
 false
 ```
 
+## ST_IsValidDetail
+
+Introduction: Returns a row, containing a boolean `valid` stating if a geometry is valid, a string `reason` stating why it is invalid and a geometry `location` pointing out where it is invalid.
```
Comment (jiayuasu) [inline docs/api/snowflake/vector-data/Function.md:1491]: This is a Snowflake table function. The doc example is not correct

File: docs/api/sql/Function.md [hunk 4]
```diff
@@ -2111,6 +2111,41 @@ Output:
...
+SQL Example:
+
+```sql
+SELECT ST_IsValidDetail(ST_GeomFromWKT('POLYGON ((30 10, 40 40, 20 40, 30 10, 10 20, 30 10))'))
+```
+
+Output:
+
+```
+{false, Ring Self-intersection at or near point (30.0, 10.0, NaN), POINT (30 10)}
```
Comment (jiayuasu) [inline docs/api/sql/Function.md:2146]: The output should be 3 columns, not a json string.

Comment (jiayuasu) [reply docs/api/sql/Function.md:2146]: Where is the flink doc example?

**Review by jiayuasu — CHANGES_REQUESTED** [review]

File: docs/api/sql/Function.md (hunk 4 above)
Comment (furqaankhan) [reply docs/api/sql/Function.md:2146]: Will make the change. This function can't be ported to Flink as Flink doesn't support struct type, you made me aware of it, hence no docs for Flink.

File: common/src/test/java/org/apache/sedona/common/FunctionsTest.java (hunk 2 above)
Comment (furqaankhan) [reply common/src/test/java/org/apache/sedona/common/FunctionsTest.java:2876]: What do you mean?

File: common/src/main/java/org/apache/sedona/common/utils/ValidDetail.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    public boolean equals(ValidDetail other) {
+        return this.valid == other.valid &&
+                (this.reason == null &&
+                this.location == null) ||
+                (this.reason.equals(other.reason) &&
+                this.location.equals(other.location));
```
Comment (Kontinuation) [inline common/src/main/java/org/apache/sedona/common/utils/ValidDetail.java:45]: Can be simplified as `Objects.equals(this.reason, other.reason) && Objects.equals(this.location, other.location)`.

File: docs/api/snowflake/vector-data/Function.md (hunk 3 above)
Comment (jiayuasu) [reply docs/api/snowflake/vector-data/Function.md:1491]: The usage of Snowflake table function is different from others. See ST_SubdivideExplode as an example: https://sedona.apache.org/1.6.0/api/snowflake/vector-data/Function/#st_subdivideexplode

```
SELECT Sedona.ST_AsText(GEOM)
FROM table(Sedona.ST_SubDivideExplode(Sedona.ST_GeomFromText('LINESTRING(0 0, 85 85, 100 100, 120 120, 21 21, 10 10, 5 5)'), 5));
```

**Review by jiayuasu — APPROVED** [review]
